### ISSUE TITLE
Add URL and Date parsing support

### ISSUE THREAD

**Issue by Kyle-Ye** [apple/swift-openapi-generator#514]
### Motivation

```
{
	"time": "2023-12-06T20:45:17.385Z",
	"url": "https://example.com"
}
```

For the above json data. The current valid openapi.yml will use `String` to represent them.

I'm hoping to somehow add support for URL and Date types directly

### Proposed solution

1. Add a variant for OpenAPI 3.1 to allow "url" and "date" for type. (Not recommended)

2. ...

### Alternatives considered

None

**Comment by czechboy0** [comment]
What @MahdiBM said is right, `Date` is already supported.

URL is not, and that's because a "URL" format is not officially supported in OpenAPI 3.0, but might be supported in OpenAPI 3.1 through JSON Schema 2020-12.

**Comment by Kyle-Ye** [comment]
> If correctly specified in `openapi.yml`:
> 
> ```yaml
> created_at:
>   type: string
>   format: date-time
> ```
> 
> The generated Swift code will use the proper `Date` type:
> 
> ```swift
> /// - Remark: Generated from `#/components/schemas/MyType/created_at`.
> package var created_at: Foundation.Date
> ```
> 
> I'm guessing you might be missing `format: date-time` or be using something that the generator can't use `Date` for.

How can I specify the date format here?

For example, I have a date value in the response

```
{ "last_posted_at": "2023-12-06T20:45:17.385Z" }
```

In the past, I use a custom JSONDecoder here.

```swift
extension DateFormatter {
    public static var discourse: DateFormatter {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'"
        formatter.timeZone = .gmt
        return formatter
    }
}

extension JSONDecoder {
    public static var discourse: JSONDecoder {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .formatted(.discourse)
        return decoder
    }
}
```

I guess now I need to config it via `Client -> configuration: Configuration -> dateTranscoder: any DateTranscoder`

**Comment by simonjbeaumont** [comment]
Yes. We provided a new date transcoder in the latest release of the runtime library that handles ISO8601 with fractional seconds, which, looking at your JSON snippet, looks to be what you might need here.

**Comment by Kyle-Ye** [comment]
> Yes. We provided a new date transcoder in the latest release of the runtime library that handles ISO8601 with fractional seconds, which, looking at your JSON snippet, looks to be what you might need here.
> 
> [swiftpackageindex.com/apple/swift-openapi-runtime/1.3.2/documentation/openapiruntime/datetranscoder/iso8601withfractionalseconds](https://swiftpackageindex.com/apple/swift-openapi-runtime/1.3.2/documentation/openapiruntime/datetranscoder/iso8601withfractionalseconds)

Thanks. I'll give it a try.

**Comment by cmcgee1024** [comment]
URL support will be helpful when interacting with the swift.org API, which often points to download artifacts.

**Comment by obrhoff** [comment]
Anything new about this feature? This is currently blocking us from migrating to swift-openapi-generator.

**Comment by czechboy0** [comment]
Can you say more about which part is blocking you?

Is it because you'd like the decoding step of the URL schema to throw an error when a string that isn't a valid URL is provided? Or because working around it by adding an extension that provides a URL view of a String property?

**Comment by obrhoff** [comment]
It’s mainly that it adds a lot of verbose code to convert strings into URLs and handle optionals. Currently, we use CreateAPI, which supports this feature.

**Comment by czechboy0** [comment]
Understood, thanks.

Yeah so this addition is mainly blocked on a proposal of the evolution story here, as we can't change the default behavior without breaking existing adopters. Once that's proposed and approved, this will unblock generating types like URL, UUID, and others.

**Comment by johankool** [comment]
I think this issue can be closed? It seems that the [evolution proposal](https://github.com/apple/swift-openapi-generator/pull/764) has been closed and the functionality is released.

The documentation for it is rather sparse.

In your `openapi-generator-config.yaml` add:

```yaml
typeOverrides:
  schemas:
    UUID: Foundation.UUID
    URL: Foundation.URL
```

Then in your spec you can use it like so:

```yaml
components:
  schemas:
          
    MyObject:
      type: object
      properties:
        uuid:
          $ref: "#/components/schemas/UUID"
        url:
          $ref: "#/components/schemas/URL"
      required:
        - text
        - url
        
    UUID:
      type: string
      format: uuid

    URL:
      type: string
      format: url
```

**Comment by czechboy0** [comment]
Possibly, though the type overrides feature is really more of an escape hatch. I still imagine we might want to generate URL out of the box in the future, so I'm also happy to keep this open.

**Comment by shadone** [comment]
Oh very nice! TIL that it is now possible to override a type!

I personally don't think it should use Foundation.URL by default as it is a bit too strict at what it considers a valid url (e.g. urls with non-latin characters without url encoding them e.g. `https://example.com/hellø`)

EDIT: unless something changed in the URL parsing? I have the following comment in my project
```
    /// Attempt to parse the provided string as URL more leniently.
    ///
    /// Some times the incoming urls are not encoded entirely correctly (as per URI spec) and Foundation's URL
    /// being very strictly spec compliant rejects those. This helper is meant to implement more lenient parsing of urls.
    ///
    /// Example of bad urls that Foundation.URL rejects:
    /// - https://matrix.to/#/#lemmy-admin-support-topics:discuss.online
    /// - https://www.reddit.com/r/oslo/comments/i63epw/lyst_til_å_finne_en_psykolog_som_hjelper_noen_med/
```

However I just re-tested this on macOS 15.6.1 and iOS 18.5 and these two urls are parsed fine by Foundation.URL
